internal/handlers/common/delete.go

```diff
@@ -15,147 +15,59 @@
 package common
 
 import (
-	"fmt"
+	"errors"
 
 	"go.uber.org/zap"
 
 	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
 	"github.com/FerretDB/FerretDB/internal/handlers/commonparams"
 	"github.com/FerretDB/FerretDB/internal/types"
-	"github.com/FerretDB/FerretDB/internal/util/must"
 )
 
-// DeleteParams represents parameters for delete operation.
+// DeleteParams represents parameters for the delete command.
 type DeleteParams struct {
-	DB, Collection string
-	Comment        string
-	Deletes        []Delete
-	Ordered        bool
+	DB         string `ferretdb:"$db"`
+	Collection string `ferretdb:"collection"`
+
+	Comment string   `ferretdb:"comment,opt"`
+	Deletes []Delete `ferretdb:"deletes,opt"`
+	Ordered bool     `ferretdb:"ordered,opt"`
+
+	Let *types.Document `ferretdb:"let,unimplemented"`
+
+	WriteConcern *types.Document `ferretdb:"writeConcern,ignored"`
 }
 
 // Delete represents single delete operation parameters.
 type Delete struct {
-	Filter  *types.Document
-	Comment string
-	Limited bool
+	Filter  *types.Document `ferretdb:"q"`
+	Limited bool            `ferretdb:"limit,numericAsBool"`
+	// TODO: https://github.com/FerretDB/FerretDB/issues/2627
+	Comment string `ferretdb:"comment,opt"`
+
+	Collation *types.Document `ferretdb:"collation,unimplemented"`
+
+	Hint string `ferretdb:"hint,ignored"`
 }
 
 // GetDeleteParams returns parameters for delete operation.
 func GetDeleteParams(document *types.Document, l *zap.Logger) (*DeleteParams, error) {
-	var err error
-
-	if err = Unimplemented(document, "let"); err != nil {
-		return nil, err
-	}
-
-	Ignored(document, l, "writeConcern")
-
-	var deletesArray *types.Array
-	if deletesArray, err = GetOptionalParam(document, "deletes", deletesArray); err != nil {
-		return nil, err
-	}
-
-	ordered := true
-	if ordered, err = GetOptionalParam(document, "ordered", ordered); err != nil {
-		return nil, err
-	}
-
-	var db, collection string
-
-	if db, err = GetRequiredParam[string](document, "$db"); err != nil {
-		return nil, err
+	params := DeleteParams{
+		Ordered: true,
 	}
 
-	collectionParam, err := document.Get(document.Command())
-	if err != nil {
-		return nil, err
-	}
-
-	var ok bool
-	if collection, ok = collectionParam.(string); !ok {
+	err := commonparams.ExtractParams(document, "delete", &params, l)
+	if errors.Is(err, commonparams.ErrFieldNotPopulated) {
```

Comment:
`"BSON field 'delete.deletes.limit' is missing but a required field"` is too specific error message since there are other field not optional. Missing field may not be `limit` but it might be `q` filter.